tests/evm/utils.py

```diff
@@ -0,0 +1,31 @@
+from zkevm_specs.evm import (
+    Opcode,
+)
+from zkevm_specs.util import linear_combine, keccak256
+
+
+# go through bytecode_source to construct byte code table of spec
+def assign_bytecode_table(bytecode_source: str, r: int):
+    bytecode = bytes.fromhex(bytecode_source)
+    byte_code_hash = linear_combine(keccak256(bytecode), r)
+    code_analysis = []
+    # code_analysis act as byte code table to holding
+    # entry(code_hash, index, push_rindex, is_code, bytecode)
+    push_x = 0
+    i = 0
+
+    # loop to analyze code flag
+    while i < len(bytecode):
+        if push_x != 0:
+            rindex = push_x
+            for x in range(0,rindex):
+                code_analysis.append((byte_code_hash, i, rindex, 0, bytecode[i]))
+                rindex -= 1
```

Comment:
don't decrease `rindex` here. it's sort of dangerous

Reply:
you also need to increase `i` inside loop, right?

Reply:
thanks for catching , fixed!

Reply:
> don't decrease `rindex` here. it's sort of dangerous
for real code, it is always zero .  for push* data segment decrease `rindex` can be conveniently constrained in state circuit.  i.e.  push3 010203 pop 
index |byte  | push_rindex|is_code  (omitting code_hash column)
  0     | 62   |        0         | 1
  1     |  01  |        3         | 0
  2    |  02   |       2         | 0
  3    |  03  |        1          | 0
  4    |  50   |       0          | 1
  when come to  push_rindex  is not zero row,  constrain  above row is push* or 'push_rindex' increase by 1, when come to push_rindex is zero row, check above row 's push_rindex is 1 or 0 definitely.  
 increasing 'push_rindex' would be more difficult to check if the push* ends or rotate more row lines.  
Right ?

Reply:
byte code table constrain would be in state circuit instead of evm circuit like bussmapping rw_table, right ?

Reply:
I mean why not change to `byte_code_table.append((byte_code_hash, i, rindex-x-1, 0, bytecode[i]))`

Just don't modify `rindex`

Reply:
that can be byte_code_table.append((byte_code_hash, i, rindex-x, 0, bytecode[i])) and updated